lib/IDE/CodeCompletion.cpp

```diff
@@ -4170,50 +4170,83 @@ class CompletionOverrideLookup : public swift::VisibleDeclConsumer {
       return nullptr;
 
     Type ResultT;
-    if (auto *FD = dyn_cast<FuncDecl>(VD))
+    if (auto *FD = dyn_cast<FuncDecl>(VD)) {
+      if (FD->getGenericParams()) {
+        // Generic function cannot have opaque result type.
+        return nullptr;
+      }
       ResultT = FD->getResultInterfaceType();
-    else if (auto *SD = dyn_cast<SubscriptDecl>(VD))
+    } else if (auto *SD = dyn_cast<SubscriptDecl>(VD)) {
+      if (SD->getGenericParams()) {
+        // Generic subscript cannot have opaque result type.
+        return nullptr;
+      }
       ResultT = SD->getElementInterfaceType();
-    else if (auto *VarD = dyn_cast<VarDecl>(VD))
+    } else if (auto *VarD = dyn_cast<VarDecl>(VD)) {
       ResultT = VarD->getInterfaceType();
-    else
-      return nullptr;
-
-    if (!ResultT->is<DependentMemberType>())
-      // The result is not associatedtype.
+    } else {
       return nullptr;
+    }
 
-    // If associatedtype doesn't have conformance/superclass constraint, we
-    // can't use opaque type.
-    auto assocTyD = ResultT->castTo<DependentMemberType>()->getAssocType();
-    if (!assocTyD->getInherited().size())
+    if (!ResultT->is<DependentMemberType>() ||
+        !ResultT->castTo<DependentMemberType>()->getAssocType())
+      // The result is not a valid associatedtype.
       return nullptr;
 
     // Try substitution to see if the associated type is resolved to concrete
     // type.
     auto substMap = currTy->getMemberSubstitutionMap(
         CurrDeclContext->getParentModule(), VD);
-    ResultT = ResultT.subst(substMap);
-    if (!ResultT || !ResultT->is<DependentMemberType>())
+    if (!ResultT.subst(substMap)->is<DependentMemberType>())
       // If resolved print it.
       return nullptr;
 
-    return assocTyD->getInherited()[0];
+    // Collect requirements on the associatedtype.
+    ProtocolDecl *protoD =
+        ResultT->castTo<DependentMemberType>()->getAssocType()->getProtocol();
+
+    SmallVector<Type, 2> opaqueTypes;
+    bool hasExplicitAnyObject = false;
+    for (auto req : protoD->getRequirementSignature()) {
```

Comment:
Instead of looking at the requirement signature, can you do this instead?

```
proto->getGenericSignature()->getConformsTo(ResultT)
proto->getGenericSignature()->getSuperclassBound(ResultT)
proto->getGenericSignature()->getLayoutConstraint(ResultT)
```

Reply:
Thanks! I wasn't aware of these methods. Updated.